### PR TITLE
Add ability to filter segments for specific dataSources on broker without creating tiers

### PR DESCRIPTION
**Pull request by nishantmonu51**
This PR adds functionality to specify "druid.broker.segment.watchedDataSources" which allows the broker to only listen for specific configured datasources. 

Changes in this PR include - 
1) Adding back FilteredServerView and related classes earlier removed in (https://github.com/druid-io/druid/commit/a32906c7fd11c9a8554df2621a172353a523a9dd)
2) Extends the changes done in https://github.com/druid-io/druid/pull/2382 to use FilteredServerView so that the filtered segments are actually removed from broker heap
3) Introduces new property "druid.broker.segment.watchedDataSources" which allows for filtering based on a whitelist set of datasources. 
4) Modifies the ClientInfoResource to use FilteredServerView and reflect the actual segments being monitored by the broker instead of full cluster view.

### CODE REVIEW

**Comment by xvrl** [comment]
👍

**Comment by fjy** [comment]
@nishantmonu51 can you explain a little bit about the problem that you want to solve with this PR? It seems like at a high level, the problem is around the Broker having to maintain too much info about segments in memory?

**Comment by xvrl** [comment]
@fjy the rationale is the same as https://github.com/druid-io/druid/pull/2382 but adding the ability to filter based on datasources in addition to tiers. In @himanshug's case tiers were used as logical groups of datasources, but in our case we use tiers mainly for replica and time-based allocation, so it didn't fit our use case.

It will also reduce memory by not storing the additional tiers / datasources metadata in PathChildrenCache, which improves memory usage even further than #2382 did.

As a result it is now possible to have small dedicated brokers for a small subset of datasources, independently of the total size of the cluster

File: server/src/main/java/io/druid/client/BatchServerInventoryView.java [hunk 1]
```diff
@@ -60,6 +72,9 @@ public BatchServerInventoryView(
         {
         }
     );
+
+    Preconditions.checkNotNull(defaultFilter);
+    this.defaultFilter = defaultFilter;
```
Comment (himanshug) [inline server/src/main/java/io/druid/client/BatchServerInventoryView.java]: nit: can just do this.defaultFilter = Preconditions.checkNotNull(defaultFilter)

File: server/src/main/java/io/druid/client/SingleServerInventoryView.java [hunk 2]
```diff
@@ -77,4 +94,58 @@ protected DruidServer removeInnerInventory(DruidServer container, String invento
...
+          }
+
+          @Override
+          public CallbackAction segmentViewInitialized()
+          {
+            return callback.segmentViewInitialized();
+          }
+        }
+    );
+  }
```
Comment (himanshug) [inline server/src/main/java/io/druid/client/SingleServerInventoryView.java]: this looks exactly same as the one in BatchServerInventoryView, can you refactor and reuse?

**Comment by himanshug** [comment]
:+1: besides minor comments

File: server/src/main/java/io/druid/client/BatchServerInventoryView.java (hunk 1 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/BatchServerInventoryView.java]: done.

File: server/src/main/java/io/druid/client/SingleServerInventoryView.java (hunk 2 above)
Comment (nishantmonu51) [inline server/src/main/java/io/druid/client/SingleServerInventoryView.java]: done.

**Comment by nishantmonu51** [comment]
@himanshug : addressed your comments.

**Pull request merged**